File: Remote/Source/Remote Elements/REView/ButtonView.m
//
// ButtonView.m
// Remote
//
// Created by Jason Cardwell on 5/24/11.
// Copyright 2011 Moondeer Studios. All rights reserved.
//
#import "RemoteElementView_Private.h"
#import "ImageView.h"
#import "TitleAttributes.h"
#import "Button.h"
#import "Command.h"
#import "ControlStateTitleSet.h"

// #define DEBUG_BV_COLOR_BG

#define MIN_HIGHLIGHT_INTERVAL 1.0
#define CORNER_RADII           CGSizeMake(5.0f, 5.0f)

MSNAMETAG_DEFINITION(REButtonViewInternal);
MSNAMETAG_DEFINITION(REButtonViewLabel);
MSNAMETAG_DEFINITION(REButtonViewActivityIndicator);

static int       ddLogLevel   = LOG_LEVEL_DEBUG;
static const int msLogContext = (LOG_CONTEXT_REMOTE | LOG_CONTEXT_FILE | LOG_CONTEXT_CONSOLE);
#pragma unused(ddLogLevel, msLogContext)


@implementation ButtonView


////////////////////////////////////////////////////////////////////////////////
#pragma mark Internal subviews and constraints
////////////////////////////////////////////////////////////////////////////////


- (void)addInternalSubviews {
  [super addInternalSubviews];

  self.subelementInteractionEnabled = NO;
  self.contentInteractionEnabled    = NO;

  _labelView = [UILabel newForAutolayout];
  [self addViewToContent:_labelView];

  _activityIndicator = [UIActivityIndicatorView newForAutolayout];
  _activityIndicator.activityIndicatorViewStyle = UIActivityIndicatorViewStyleWhiteLarge;
  _activityIndicator.color                      = defaultTitleHighlightColor();
  [self addViewToOverlay:_activityIndicator];
}

- (void)updateConstraints {
  [super updateConstraints];

  NSString * labelNametag    = ClassNametagWithSuffix(@"InternalLabel");
  NSString * activityNametag = ClassNametagWithSuffix(@"InternalActivity");

  if (![self constraintsWithNametagPrefix:labelNametag]) {
    UIEdgeInsets titleInsets = self.model.titleEdgeInsets;
    NSString   * constraints =
      $(@"'%1$@-Label$@' _labelView.left = self.left + %3$f @900\n"
        "'%1$@-Label$@' _labelView.top = self.top + %4$f @900\n"
        "'%1$@-Label$@' _labelView.bottom = self.bottom - %5$f @900\n"
        "'%1$@-Label$@' _labelView.right = self.right - %6$f @900\n"
        "'%2$@-Activity$@' _activityIndicator.centerX = self.centerX\n"
        "'%2$@-Activity$@' _activityIndicator.centerY = self.centerY",
        labelNametag, activityNametag, titleInsets.left, titleInsets.top, titleInsets.bottom, titleInsets.right);

    NSDictionary * views = NSDictionaryOfVariableBindings(self, _labelView, _activityIndicator);

    [self addConstraints:[NSLayoutConstraint constraintsByParsingString:constraints views:views]];
  }
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark Gestures
////////////////////////////////////////////////////////////////////////////////


- (void)attachGestureRecognizers {
  [super attachGestureRecognizers];

  _longPressGesture = [MSLongPressGestureRecognizer gestureWithTarget:self
                                                               action:@selector(handleLongPress:)];
  _longPressGesture.delaysTouchesBegan = NO;
  _longPressGesture.delegate           = self;
  [self addGestureRecognizer:_longPressGesture];

  _tapGesture                         = [UITapGestureRecognizer gestureWithTarget:self action:@selector(handleTap:)];
  _tapGesture.numberOfTapsRequired    = 1;
  _tapGesture.numberOfTouchesRequired = 1;
  _tapGesture.delaysTouchesBegan      = NO;
  _tapGesture.delegate                = self;
  [self addGestureRecognizer:_tapGesture];
}

/**
   Single tap action executes the primary button command
 */
- (void)handleTap:(UITapGestureRecognizer *)gestureRecognizer {
  if (gestureRecognizer.state == UIGestureRecognizerStateRecognized) {
    self.highlighted = YES;
    assert(self.model.highlighted);

    MSDelayedRunOnMain(_options.minHighlightInterval,
                       ^{
      _flags.highlightActionQueued = NO;
      self.highlighted = NO;
      [self setNeedsDisplay];
    });

    REActionHandler handler = _actionHandlers[@(RESingleTapAction)];

    if (handler)
      handler();

    else
      [self buttonActionWithOptions:CommandOptionDefault];
  } else if (gestureRecognizer.state == UIGestureRecognizerStateCancelled && !_flags.longPressActive) {
//        self.highlighted = NO;
//        [self setNeedsDisplay];
  }
}

/**
   Long press action executes the secondary button command
 */
- (void)handleLongPress:(MSLongPressGestureRecognizer *)gestureRecognizer {
  if (gestureRecognizer.state == UIGestureRecognizerStateBegan) {
    REActionHandler handler = _actionHandlers[@(RELongPressAction)];

    if (handler)
      handler();

    else
      [self buttonActionWithOptions:CommandOptionLongPress];
  } else if (gestureRecognizer.state == UIGestureRecognizerStatePossible) {
    _flags.longPressActive = YES;
    self.highlighted       = YES;
    [self setNeedsDisplay];
  }
}

/**
   Enables or disables tap and long press gestures
 */
- (void)updateGesturesEnabled:(BOOL)enabled {
  _tapGesture.enabled       = enabled;
  _longPressGesture.enabled = enabled;
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark ￼Button state
////////////////////////////////////////////////////////////////////////////////


- (void)updateState {
  UIControlState currentState = self.state;

  self.userInteractionEnabled = ((currentState & UIControlStateDisabled) ? NO : YES);

  BOOL invalidate = NO;

  NSAttributedString * title = self.model.title;

  if (![_labelView.attributedText isEqualToAttributedString:title]) {
    _labelView.attributedText = title;
    invalidate                = YES;
  }

  UIImage * icon = self.model.icon.colorImage;

  if (_icon != icon) {
    _icon      = icon;
    invalidate = YES;
  }

  self.backgroundColor = self.model.backgroundColor;

  if (invalidate) {
    [self invalidateIntrinsicContentSize];
    [self setNeedsDisplay];
  }

}

- (UIControlState)state { return (UIControlState)self.model.state; }


////////////////////////////////////////////////////////////////////////////////
#pragma mark Button actions
////////////////////////////////////////////////////////////////////////////////


- (void)setActionHandler:(REActionHandler)handler forAction:(REAction)action {
  _actionHandlers[@(action)] = handler;
}

- (void)buttonActionWithOptions:(CommandOptions)options {
  assert(self.model);

  if (!self.editing && _flags.commandsActive) {
    if (_flags.longPressActive) {
      _flags.longPressActive = NO;
      [self setNeedsDisplay];
    }

    if (_flags.activityIndicator) [_activityIndicator startAnimating];

    CommandCompletionHandler completion =
      ^(BOOL success, NSError * error)
    {
      if ([_activityIndicator isAnimating])
        MSRunAsyncOnMain(^{ [_activityIndicator stopAnimating]; });
    };

    [self.model executeCommandWithOptions:options completion:completion];
  }
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark Content size
////////////////////////////////////////////////////////////////////////////////


- (CGSize)intrinsicContentSize { return self.minimumSize; }

- (CGSize)minimumSize {
  CGRect frame = (CGRect) {
    .size = REMinimumSize
  };

  NSMutableSet * titles = [NSMutableSet set];

  for (NSString *mode in self.model.modes) {
    ControlStateTitleSet * titleSet = [self.model titlesForMode:mode];
    if (titleSet) [titles addObjectsFromArray:[titleSet allValues]];
  }

  if ([titles count]) {

    CGFloat maxWidth = 0.0, maxHeight = 0.0;

    for (NSAttributedString * title in titles) {

      CGSize titleSize = [title size];
      UIEdgeInsets titleInsets = self.titleEdgeInsets;

      titleSize.width  += titleInsets.left + titleInsets.right;
      titleSize.height += titleInsets.top + titleInsets.bottom;

      maxWidth = MAX(titleSize.width, maxWidth);
      maxHeight = MAX(titleSize.height, maxHeight);

    }

    frame.size.width = MAX(maxWidth, frame.size.width);
    frame.size.height = MAX(maxHeight, frame.size.height);
  }


/*
   NSAttributedString * title = self.model.title;

  if (title) {
    CGSize       titleSize   = [title size];
    UIEdgeInsets titleInsets = self.titleEdgeInsets;

    titleSize.width  += titleInsets.left + titleInsets.right;
    titleSize.height += titleInsets.top + titleInsets.bottom;
    frame             = CGRectUnion(frame, (CGRect) {.size = titleSize });
  }

  if (_icon) {
    CGSize       iconSize    = [_icon size];
    UIEdgeInsets imageInsets = self.imageEdgeInsets;

    iconSize.width  += imageInsets.left + imageInsets.right;
    iconSize.height += imageInsets.top + imageInsets.bottom;
    frame            = CGRectUnion(frame, (CGRect) {.size = iconSize });
  }

  UIEdgeInsets contentInsets = self.contentEdgeInsets;

  frame.size.width  += contentInsets.left + contentInsets.right;
  frame.size.height += contentInsets.top + contentInsets.bottom;

*/

  if (self.model.proportionLock && !CGSizeEqualToSize(self.bounds.size, CGSizeZero)) {

    CGSize currentSize = self.bounds.size;

    if (currentSize.width > currentSize.height)
      frame.size.height = (frame.size.width * currentSize.height) / currentSize.width;

    else
      frame.size.width = (frame.size.height * currentSize.width) / currentSize.height;
  }

  return frame.size;
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark Subelement views
////////////////////////////////////////////////////////////////////////////////


- (void)addSubelementView:(RemoteElementView *)view {}

- (void)removeSubelementView:(RemoteElementView *)view {}

- (void)addSubelementViews:(NSSet *)views {}

- (void)removeSubelementViews:(NSSet *)views {}

- (NSArray *)subelementViews { return nil; }


////////////////////////////////////////////////////////////////////////////////
#pragma mark Initialization
////////////////////////////////////////////////////////////////////////////////


- (void)initializeIVARs {
  _actionHandlers               = [@{} mutableCopy];
  self.cornerRadii              = CORNER_RADII;
  _options.minHighlightInterval = MIN_HIGHLIGHT_INTERVAL;
  _flags.commandsActive         = YES;

  [super initializeIVARs];
}

- (void)initializeViewFromModel {
  [super initializeViewFromModel];

  _longPressGesture.enabled = (self.model.longPressCommand != nil);
  _flags.activityIndicator  = self.model.command.indicator;

  [self updateState];
  [self invalidateIntrinsicContentSize];
  [self setNeedsDisplay];
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark Key-value observing
////////////////////////////////////////////////////////////////////////////////


- (MSDictionary *)kvoRegistration {

  MSDictionary * reg = [super kvoRegistration];

    reg[@"selected"] = ^(MSKVOReceptionist * receptionist) {
      ButtonView * buttonView = (__bridge ButtonView *)receptionist.context;
      [(__bridge ButtonView *)receptionist.context updateState];
    };

    reg[@"enabled"] = ^(MSKVOReceptionist * receptionist) {
      ButtonView * buttonView = (__bridge ButtonView *)receptionist.context;
      BOOL         enabled    = [receptionist.change[NSKeyValueChangeNewKey] boolValue];
      buttonView.enabled = enabled;
    };

    reg[@"highlighted"] = ^(MSKVOReceptionist * receptionist) {
      ButtonView * buttonView = (__bridge ButtonView *)receptionist.context;
      [(__bridge ButtonView *)receptionist.context updateState];
    };

    reg[@"command"] = ^(MSKVOReceptionist * receptionist) {
      ButtonView * buttonView = (__bridge ButtonView *)receptionist.context;
      buttonView->_flags.activityIndicator = buttonView.model.command.indicator;
    };

    reg[@"style"] = ^(MSKVOReceptionist * receptionist) {
      ButtonView * buttonView = (__bridge ButtonView *)receptionist.context;
      [buttonView setNeedsDisplay];
    };

    reg[@"title"] = ^(MSKVOReceptionist * receptionist) {
      ButtonView         * buttonView = (__bridge ButtonView *)receptionist.context;
      NSAttributedString * title      = NilSafe(receptionist.change[NSKeyValueChangeNewKey]);
      buttonView->_labelView.attributedText = title;
    };

    reg[@"image"] = ^(MSKVOReceptionist * receptionist) {
      [(__bridge ButtonView *)receptionist.context updateState];
    };

    reg[@"icon"] = ^(MSKVOReceptionist * receptionist) {
      ButtonView * buttonView = (__bridge ButtonView *)receptionist.context;
      ImageView  * icon       = NilSafe(receptionist.change[NSKeyValueChangeNewKey]);
      buttonView->_icon = icon.colorImage;
      [buttonView setNeedsDisplay];
    };

  return reg;
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark Editing
////////////////////////////////////////////////////////////////////////////////


- (void)setEditingMode:(REEditingMode)editingMode {
  [super setEditingMode:editingMode];
  _flags.commandsActive = (editingMode == REEditingModeNotEditing) ? YES : NO;
  [self updateGesturesEnabled:_flags.commandsActive];
}

////////////////////////////////////////////////////////////////////////////////
#pragma mark Drawing
////////////////////////////////////////////////////////////////////////////////


- (void)drawContentInContext:(CGContextRef)ctx inRect:(CGRect)rect {
  if (_icon) {
    UIGraphicsPushContext(ctx);
    CGRect insetRect = UIEdgeInsetsInsetRect(self.bounds, self.imageEdgeInsets);
    CGSize imageSize = (CGSizeContainsSize(insetRect.size, _icon.size)
                        ? _icon.size
                        : CGSizeAspectMappedToSize(_icon.size, insetRect.size, YES));
    CGRect imageRect = CGRectMake(CGRectGetMidX(insetRect) - imageSize.width / 2.0,
                                  CGRectGetMidY(insetRect) - imageSize.height / 2.0,
                                  imageSize.width,
                                  imageSize.height);

    if (_options.antialiasIcon) {
      CGContextSetAllowsAntialiasing(ctx, YES);
      CGContextSetShouldAntialias(ctx, YES);
    }

    [_icon drawInRect:imageRect];
    UIGraphicsPopContext();
  }
}

@end

File: UI/Source/ButtonView.swift
//
//  ButtonView.swift
//  Remote
//
//  Created by Jason Cardwell on 11/07/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import UIKit
import MoonKit
import DataModel

public class ButtonView: RemoteElementView {

  var button: Button { return model as! Button }

  // MARK: - Cached model values

  public private(set) var title: NSAttributedString? { didSet { invalidateIntrinsicContentSize(); setNeedsDisplay() } }
  public private(set) var icon: UIImage? { didSet { invalidateIntrinsicContentSize(); setNeedsDisplay() } }
  public private(set) var foregroundColor: UIColor? { didSet { setNeedsDisplay() } }
  private var _backgroundColor: UIColor? { didSet { setNeedsDisplay() } }
  public override var backgroundColor: UIColor? { get { return _backgroundColor } set { _backgroundColor = newValue } }

  // MARK: - Actions

  var tapAction: ((Void) -> Void)?
  var pressAction: ((Void) -> Void)?

	/**
	executeActionWithOption:

	:param: option CommandOption
	*/
	func executeActionWithOption(option: Command.Option) {
		if !isEditing {
			if button.command != nil && button.command!.indicator { activityIndicator.startAnimating() }
      MSLogDebug("executing command for button named '\(button.name)'")
			button.executeCommandWithOption(option) {
				(success: Bool, error: NSError?) -> Void in
					if self.activityIndicator.isAnimating() {
						NSOperationQueue.mainQueue().addOperationWithBlock {
							self.activityIndicator.stopAnimating()
						}
					}
			}
		}
	}

  // MARK: - Gestures

  weak var tapGesture :UITapGestureRecognizer!
  weak var longPressGesture: UILongPressGestureRecognizer!

  /** attachGestureRecognizers */
	override func attachGestureRecognizers() {
		super.attachGestureRecognizers()

		let longPressGesture = UILongPressGestureRecognizer(target: self, action: "handleLongPress:")
		longPressGesture.delaysTouchesBegan = false
		addGestureRecognizer(longPressGesture)
		self.longPressGesture = longPressGesture

		let tapGesture = UITapGestureRecognizer(target: self, action: "handleTap:")
		tapGesture.numberOfTapsRequired    = 1
		tapGesture.numberOfTouchesRequired = 1
		tapGesture.delaysTouchesBegan      = false
		addGestureRecognizer(tapGesture)
		self.tapGesture = tapGesture
	}

	/**
	Single tap action executes the primary button command

	:param: gestureRecognizer UITapGestureRecognizer
	*/
	func handleTap(gestureRecognizer: UITapGestureRecognizer) {
    MSLogDebug("tapped button '\(button.name)'")
		if gestureRecognizer.state == .Ended {
			button.highlighted = true
			MSDelayedRunOnMain(1, {self.button.highlighted = false})
			tapAction?() ?? executeActionWithOption(.Default)
		}
	}

	/**
	Long press action executes the secondary button command

	:param: gestureRecognizer UILongPressGestureRecognizer
	*/
	func handleLongPress(gestureRecognizer: UILongPressGestureRecognizer) {
    MSLogDebug("pressed button '\(button.name)'")
		if gestureRecognizer.state == .Ended {
			pressAction?() ?? executeActionWithOption(.LongPress)

		} else if gestureRecognizer.state == .Possible {
			button.highlighted = true
			setNeedsDisplay()
		}
	}

  // MARK: - Internal views

  /*weak var labelView: UILabel!*/
  weak var activityIndicator: UIActivityIndicatorView!

	/** addInternalSubviews */
	override func addInternalSubviews() {
		super.addInternalSubviews()

		let activityIndicator = UIActivityIndicatorView.newForAutolayout()
		activityIndicator.activityIndicatorViewStyle = .WhiteLarge
		activityIndicator.color = UIColor.whiteColor()
		addSubview(activityIndicator)
		self.activityIndicator = activityIndicator
	}

  // MARK: - Constraints

	/** updateConstraints */
	override public func updateConstraints() {
		super.updateConstraints()

    let identifier = createIdentifier(self, ["Button", "Internal"])
    if constraintsWithIdentifier(identifier).count == 0 {
      constrain(identifier: identifier, activityIndicator.centerX => self.centerX, activityIndicator.centerY => self.centerY)
    }
	}

  // MARK: - Size

 /**
	intrinsicContentSize

	:returns: CGSize
	*/
	override public func intrinsicContentSize() -> CGSize { return minimumSize }

	override public var minimumSize: CGSize {
		var frame = CGRect(size: RemoteElementView.MinimumSize)
    if let title = button.title {
      var titleSize = title.size()
      let titleInsets = button.titleEdgeInsets
      titleSize.width += titleInsets.left + titleInsets.right
      titleSize.height += titleInsets.top + titleInsets.bottom
      frame.size.width = max(titleSize.width, frame.size.width)
      frame.size.height = max(titleSize.height, frame.size.height)
    }

	  if model.constraintManager.proportionLock && bounds.size != CGSize.zeroSize {
	  	let currentSize = bounds.size

	    if currentSize.width > currentSize.height {
	    	frame.size.height = (frame.size.width * currentSize.height) / currentSize.width
    	} else {
    		frame.size.width = (frame.size.height * currentSize.width) / currentSize.height
    	}
    }

	  return frame.size
	}

  // MARK: - KVO

	/**
	kvoRegistration

	:returns: [Property:KVOReceptionist.Observation]
	*/
	override func kvoRegistration() -> [Property:KVOReceptionist.Observation] {
		var registry = super.kvoRegistration()
    registry["background"] = nil
    registry["backgroundColor"] = {
      RemoteElementView.dumpObservation($0)
      ($0.observer as? ButtonView)?.backgroundColor = ($0.object as? Button)?.backgroundColor
    }
    registry["foregroundColor"] = {
      RemoteElementView.dumpObservation($0)
      ($0.observer as? ButtonView)?.foregroundColor = ($0.object as? Button)?.foregroundColor
    }
    registry["title"] = {
      RemoteElementView.dumpObservation($0)
      ($0.observer as? ButtonView)?.title = ($0.object as? Button)?.title
    }
		registry["icon"] = {
      RemoteElementView.dumpObservation($0)
      ($0.observer as? ButtonView)?.icon = ($0.object as? Button)?.icon?.colorImage
    }
    registry["longPressCommand"] = {
      RemoteElementView.dumpObservation($0)
      ($0.observer as? ButtonView)?.longPressGesture.enabled = ($0.object as? Button)?.longPressCommand != nil
    }
    registry["state"] = {
      RemoteElementView.dumpObservation($0)
      ($0.observer as? ButtonView)?.updateStateSensitiveProperties()
    }
		return registry
	}

  // MARK: - Updating

  /** updateViewFromModel */
  override func updateViewFromModel() {
    super.updateViewFromModel()
    longPressGesture.enabled = button.longPressCommand != nil
    updateStateSensitiveProperties()
  }

  /** updateStateSensitiveProperties */
  func updateStateSensitiveProperties() {
    title = button.title
    icon  = button.icon?.colorImage
    backgroundColor = button.backgroundColor
    foregroundColor = button.foregroundColor
    invalidateIntrinsicContentSize()
  }

  // MARK: - Editing

	override public var editingMode: RemoteElement.BaseType {
		didSet {
			tapGesture.enabled = !isEditing
			longPressGesture.enabled = !isEditing
		}
	}

  // MARK: - Drawing

	/**
	drawRect:

	:param: rect CGRect
	*/
	override public func drawRect(rect: CGRect) {
    if hasOption(.DrawBackground, button.style) { drawWithBackgroundInRect(rect) } else  { drawWithoutBackgroundInRect(rect) }
	}

  /**
  drawWithoutBackgroundInRect:

  :param: rect CGRect
  */
  private func drawWithoutBackgroundInRect(rect: CGRect) {
    let context = UIGraphicsGetCurrentContext()
    let fgColor = foregroundColor ?? Painter.defaultForegroundColor
    let shadow: NSShadow? = button.highlighted ? NSShadow(color: fgColor, offset: CGSize.zeroSize, blurRadius: 5) : nil

    CGContextSaveGState(context)
    shadow?.setShadow()
    CGContextBeginTransparencyLayer(context, nil)

    CGContextSaveGState(context)
    fgColor.setFill()
    UIRectFill(rect)

    CGContextSetBlendMode(context, kCGBlendModeDestinationIn)
    CGContextBeginTransparencyLayer(context, nil)

    if let image = icon {
      let imageAttrs = Painter.Attributes(
        rect: (button.contentEdgeInsets + button.imageEdgeInsets).insetRect(bounds),
        foregroundColor: fgColor,
        accentShadow: shadow
      )
      Painter.drawImage(image, withAttributes: imageAttrs, boundByShape: button.shape)
    }


    if let attributedText = title where attributedText.length > 0 {
      let txtAttrs = Painter.Attributes(
        rect: (button.contentEdgeInsets + button.titleEdgeInsets).insetRect(bounds),
        shadow: shadow,
        text: attributedText.string,
        fontAttributes: attributedText.attributesAtIndex(0, effectiveRange: nil)
      )
      Painter.drawText(attributedText.string, withAttributes: txtAttrs, boundByShape: button.shape)
    }

    CGContextEndTransparencyLayer(context)
    CGContextRestoreGState(context)

    CGContextEndTransparencyLayer(context)
    CGContextRestoreGState(context)

  }

  /**
  drawWithBackgroundInRect:

  :param: rect CGRect
  */
  private func drawWithBackgroundInRect(rect: CGRect) {

    if button.shape == .Undefined { return }

    let context = UIGraphicsGetCurrentContext()

    let baseRect = rect.integerRect
    let bleedRect = baseRect.rectByInsetting(dx: 4, dy: 4)

    let bgColor = backgroundColor ?? Painter.defaultBackgroundColor
    let fgColor = foregroundColor ?? Painter.defaultForegroundColor
    let accentShadow: NSShadow? = button.highlighted ? NSShadow(color: fgColor, offset: CGSize.zeroSize, blurRadius: 5) : nil

    // Draw shape filled with accent color
    let bleedThroughAttrs = Painter.Attributes(rect: bleedRect, color: fgColor)
    Painter.drawShape(button.shape, withAttributes: bleedThroughAttrs)

    CGContextSaveGState(context)                                                            // context: •
    accentShadow?.setShadow()
    CGContextBeginTransparencyLayer(context, nil)                                           // transparency: •

    let baseAttrs = Painter.Attributes(
      rect: baseRect,
      color: bgColor,
      accentColor: fgColor,
      foregroundColor: fgColor
    )
    Painter.drawBaseWithShape(button.shape, attributes: baseAttrs)

    CGContextSaveGState(context)                                                            // context: ••
    CGContextSetBlendMode(context, kCGBlendModeDestinationOut)
    accentShadow?.setShadow()
    CGContextBeginTransparencyLayer(context, nil)                                           // transparency: ••

    if let image = icon {

      let imageAttrs = Painter.Attributes(
        rect: bleedRect,
        color: fgColor,
        accentShadow: accentShadow
      )
      Painter.drawImage(image, withAttributes: imageAttrs, boundByShape: button.shape)
    }

    if let attributedText = title where attributedText.length > 0 {

      let txtAttrs = Painter.Attributes(
        rect: baseRect,
        shadow: accentShadow,
        text: attributedText.string,
        fontAttributes: attributedText.attributesAtIndex(0, effectiveRange: nil)
      )
      Painter.drawText(attributedText.string, withAttributes: txtAttrs, boundByShape: button.shape)
    }

    CGContextEndTransparencyLayer(context)                                                  // transparency: •
    CGContextRestoreGState(context)                                                         // context: •

    CGContextEndTransparencyLayer(context)                                                  // transparency:
    CGContextRestoreGState(context)                                                         // context:

    if hasOption(.ApplyGloss, button.style) {
      let glossAttrs = Painter.Attributes(rect: baseRect, alpha: 0.15, blendMode: kCGBlendModeSoftLight)
      Painter.drawGlossWithShape(button.shape, attributes: glossAttrs)
    }
  }

}
